src/include/platform/BuildTime.h

```diff
@@ -0,0 +1,71 @@
+#define BUILD_MONTH_IS_MAY(_date) ((_date)[0] == 'M' && (_date)[1] == 'a' && (_date)[2] == 'y')
+#define BUILD_MONTH_IS_JUN(_date) ((_date)[0] == 'J' && (_date)[1] == 'u' && (_date)[2] == 'n')
+#define BUILD_MONTH_IS_JUL(_date) ((_date)[0] == 'J' && (_date)[1] == 'u' && (_date)[2] == 'l')
+#define BUILD_MONTH_IS_AUG(_date) ((_date)[0] == 'A' && (_date)[1] == 'u')
+#define BUILD_MONTH_IS_SEP(_date) ((_date)[0] == 'S')
+#define BUILD_MONTH_IS_OCT(_date) ((_date)[0] == 'O')
+#define BUILD_MONTH_IS_NOV(_date) ((_date)[0] == 'N')
+#define BUILD_MONTH_IS_DEC(_date) ((_date)[0] == 'D')
+
+#define COMPUTE_BUILD_MONTH(_date)                                                                                                 \
+    ((BUILD_MONTH_IS_JAN(_date))                                                                                                   \
+         ? 1                                                                                                                       \
+         : (BUILD_MONTH_IS_FEB(_date)) ? 2                                                                                         \
+                                       : (BUILD_MONTH_IS_MAR(_date)) ? 3                                                           \
+                                                                     : (BUILD_MONTH_IS_APR(_date)) ? 4                             \
+                                                                                                   : (BUILD_MONTH_IS_MAY(_date))   \
+                         ? 5                                                                                                       \
+                         : (BUILD_MONTH_IS_JUN(_date)) ? 6                                                                         \
+                                                       : (BUILD_MONTH_IS_JUL(_date)) ? 7                                           \
+                                                                                     : (BUILD_MONTH_IS_AUG(_date))                 \
+                                     ? 8                                                                                           \
+                                     : (BUILD_MONTH_IS_SEP(_date)) ? 9                                                             \
+                                                                   : (BUILD_MONTH_IS_OCT(_date)) ? 10                              \
+                                                                                                 : (BUILD_MONTH_IS_NOV(_date))     \
+                                                 ? 11                                                                              \
+                                                 : (BUILD_MONTH_IS_DEC(_date)) ? 12 : /* error default */ 99)
+
+#define COMPUTE_BUILD_HOUR(_time) (((_time)[0] - '0') * 10 + (_time)[1] - '0')
```

Comment:
Isn't using __TIME__ macro causing that incremental builds are no longer efficient? I'm concerned by how this can impact development. Maybe we could instead generate this build time in GN so that it is run once per `gn gen`?

Reply:
These are being routed through the ConfugrationManager by passing in our preexisting `#define CHIP_DEVICE_CONFIG_FIRMWARE_BUILD_DATE __DATE__`, `#define CHIP_DEVICE_CONFIG_FIRMWARE_BUILD_TIME __TIME__` configs, so at the very least only one compilation unit would ever need to be rebuilt.  In practice and on the platforms I've tried however, this actually doesn't trigger rebuild even for ConfigManager.  So no, incremental builds are not losing efficiency so far as I've seen.

I think the bigger concern would be non-reproducible builds.  To be a good citizen about this, I'm looking at the problem this way:

* the spec REQUIRES that firmware know its build time, so we need _something_
* I also take the preexistence of the `CHIP_DEVICE_CONFIG_FIRMWARE_BUILD_DATE` and `CHIP_DEVICE_CONFIG_FIRMWARE_BUILD_TIME`, which already referenced `__DATE__` and `__TIME__`, to mean we have decided these can be acceptable, at least in an example application
* but hiding `__DATE__` and `__TIME__` behind our CONFIGs still allows for injecting the values with something static to allow for reproducible builds
* also, the routing through the ConfigurationManager and provision of an in-memory Setter allows for implementations that can't have these in program memory at all -> if one needs their implementation to get these values from some other arbitrary source, it can be done without forking

Let me know if this addresses your concern.  I did think quite a bit about the implications of using `__DATE__` and `__TIME__.

Reply:
Yes, thanks for explanation.